dask/highlevelgraph.py

```diff
@@ -24,6 +27,122 @@ def _find_layer_containing_key(key):
     return ret
 
 
+class Layer(Mapping):
+    """High level graph layer
+
+    This abstract class establish a protocol for high level graph layers.
+    """
+
+    def cull(self, keys: Set) -> "Layer":
+        """Return a new Layer with only the tasks required to calculate `keys`.
+
+        In other words, remove unnecessary tasks from the layer.
+
+        Examples
+        --------
+        >>> d = Layer({'x': 1, 'y': (inc, 'x'), 'out': (add, 'x', 10)})  # doctest: +SKIP
+        >>> d.cull({'out'})  # doctest: +SKIP
+        {'x': 1, 'out': (add, 'x', 10)}
+
+        Returns
+        -------
+        layer: Layer
+            Culled layer
+        """
+
+        seen = set()
+        out = {}
+        work = set(flatten(keys))
+        while work:
+            new_work = set()
+            for k in work:
+                out[k] = self[k]
+                for d in get_dependencies(self, k):
+                    if d not in seen:
+                        seen.add(d)
+                        new_work.add(d)
+            work = new_work
+
+        return BasicLayer(out)
+
+    def get_external_dependencies(self, all_hlg_keys: Set) -> Set:
+        """Get external dependencies
+
+        Parameters
+        ----------
+        all_hlg_keys : set
+            Set of all keys in the high level graph.
+
+        Returns
+        -------
+        deps: set
+            Set of dependencies
+        """
+
+        all_deps = keys_in_tasks(all_hlg_keys, self.values())
+        return all_deps.difference(self.keys())
+
+
+class BasicLayer(Layer):
+    """Basic implementation of `Layer`
+
+    Parameters
+    ----------
+    mapping : Mapping
+        The mapping between keys and tasks, typically a dask graph.
+    key_dependencies : Mapping[Hashable, Set], optional
+        Mapping between keys and their dependencies
+    unstructured_tasks : Mapping[Hashable, Set], optional
+        Mapping between keys and unstructured tasks, which are tasks
+        that have no dependencies information.
```

Comment:
I don't yet understand the objective of this state.  Are you saying that these are tasks without dependencies?  Or is this for tasks that have dependencies, but for some reason we haven't captured those dependencies.  

If the latter, when does this occur?

Reply:
It is the latter case. This occur in blockwise for [non-blockwise tasks](https://github.com/dask/dask/blob/184134deac75a93ff937d1ca4804f2bdbeaa8784/dask/blockwise.py#L457), which contains keys that are constant throughout the blockwise operation. At the execution of [`make_blockwise_graph()`](https://github.com/dask/dask/blob/184134deac75a93ff937d1ca4804f2bdbeaa8784/dask/blockwise.py#L216) we don't have a set of all HLG keys thus we have to wait until `get_external_dependencies()` to identify the keys in `unstructured_tasks`.

Reply:
@mrocklin, I have changed this to 
```python 
possible_dependencies : Mapping[Hashable, Set], optional
        Mapping between keys and possible dependencies, which are dependencies
        that might also be literals.
```

Reply:
Ah, so this would occur for example if I add a dask array to a dask delayed object?

Reply:
Yes, every time we have keys in `kwargs` or in `None`-indexed inputs to  [`make_blockwise_graph()`](https://github.com/dask/dask/blob/master/dask/blockwise.py#L247)

Reply:
Actually think about this some more, I realized that because its non-blockwise keys we can optimize this by only having a set of global dependencies.